### PR TITLE
Update kube Infra resources if they already exist

### PR DESCRIPTION
**Pull request by arkodg**
Fixes: https://github.com/envoyproxy/gateway/issues/341

Signed-off-by: Arko Dasgupta <arko@tetrate.io>

### CODE REVIEW

File: internal/infrastructure/kubernetes/deployment.go [hunk 1]
```diff
@@ -248,22 +208,31 @@ func expectedContainers(infra *ir.Infra) ([]corev1.Container, error) {
...
 	}
 
-	if err := i.Client.Create(ctx, expected); err != nil {
+	if err := i.Client.Create(ctx, deploy); err != nil {
 		if kerrors.IsAlreadyExists(err) {
-			return expected, nil
+			// Update deployment if it exists.
+			if err := i.Client.Update(ctx, deploy); err != nil {
+				return fmt.Errorf("failed to update deployment %s/%s: %w",
+					deploy.Namespace, deploy.Name, err)
```
Comment (danehans) [inline internal/infrastructure/kubernetes/deployment.go:223]: IMO we should compare expected with actual, e.g. what exists in the controller-runtime cache, and only update if managed fields, e.g. `spec`, change. I'm fine proceeding with the approach in this PR for v0.2.0 as long as a tracking issue is created for doing the comparison.

File: internal/infrastructure/kubernetes/service.go [hunk 2]
```diff
@@ -101,20 +61,29 @@ func (i *Infra) expectedService(infra *ir.Infra) *corev1.Service {
...
+	svc := i.expectedService(infra)
+	err := i.Client.Create(ctx, svc)
 	if err != nil {
 		if kerrors.IsAlreadyExists(err) {
-			return expected, nil
+			// Update service if its exists
+			if err := i.Client.Update(ctx, svc); err != nil {
+				return fmt.Errorf("failed to update service %s/%s: %w",
+					svc.Namespace, svc.Name, err)
+			}
```
Comment (danehans) [inline internal/infrastructure/kubernetes/service.go:75]: IMO we should compare expected with actual, e.g. what exists in the controller-runtime cache, and only update if managed fields, e.g. `spec`, change. I'm fine proceeding with the approach in this PR for v0.2.0 as long as a tracking issue is created for doing the comparison.

File: internal/infrastructure/kubernetes/serviceaccount.go [hunk 3]
```diff
@@ -17,78 +15,43 @@ const (
...
+	svcAccount := i.expectedServiceAccount()
+	err := i.Client.Create(ctx, svcAccount)
 	if err != nil {
 		if kerrors.IsAlreadyExists(err) {
-			return expected, nil
+			// Update service account if it does not exist
+			if err := i.Client.Update(ctx, svcAccount); err != nil {
+				return fmt.Errorf("failed to update serviceaccount %s/%s: %w",
+					svcAccount.Namespace, svcAccount.Name, err)
+			}
```
Comment (danehans) [inline internal/infrastructure/kubernetes/serviceaccount.go:43]: IMO we should compare expected with actual, e.g. what exists in the controller-runtime cache, and only update if managed fields, e.g. `spec`, change. I'm fine proceeding with the approach in this PR for v0.2.0 as long as a tracking issue is created for doing the comparison.

File: internal/infrastructure/kubernetes/deployment.go (hunk 1 above)
Comment (arkodg) [reply internal/infrastructure/kubernetes/deployment.go]: addressed your feedback in this PR

**Comment by arkodg** [comment]
since infra service ports are hardcoded, I prefer to add update tests for service in https://github.com/envoyproxy/gateway/pull/302

File: internal/infrastructure/kubernetes/serviceaccount_test.go [hunk 4]
```diff
@@ -93,8 +90,10 @@ func TestCreateOrUpdateServiceAccount(t *testing.T) {
 			}
 			err := kube.createOrUpdateServiceAccount(context.Background(), tc.in)
 			require.NoError(t, err)
-			opts := cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ResourceVersion")
-			assert.Equal(t, true, cmp.Equal(tc.out.ServiceAccount, kube.Resources.ServiceAccount, opts))
+			// Clear ResourceVersion before the equality test.
+			tc.out.ServiceAccount.ResourceVersion = ""
+			kube.Resources.ServiceAccount.ResourceVersion = ""
+			require.Equal(t, tc.out.ServiceAccount, kube.Resources.ServiceAccount)
```
Comment (LukeShu) [inline internal/infrastructure/kubernetes/serviceaccount_test.go:96]: 1) The commit message on this change was truncated beyond usefulness, the full trace is 
```
==================
WARNING: DATA RACE
Read at 0x00c0005b9930 by goroutine 40:
  k8s.io/apimachinery/pkg/apis/meta/v1.(*ObjectMeta).GetResourceVersion()
      /home/lukeshu/src/github.com/envoyproxy/gateway/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:145 +0x3e
  k8s.io/api/core/v1.(*ServiceAccount).GetResourceVersion()
      <autogenerated>:1 +0x1d
  sigs.k8s.io/controller-runtime/pkg/client/fake.versionedTracker.Add()
      /home/lukeshu/src/github.com/envoyproxy/gateway/vendor/sigs.k8s.io/controller-runtime/pkg/client/fake/client.go:195 +0x1be
  sigs.k8s.io/controller-runtime/pkg/client/fake.(*ClientBuilder).Build()
      /home/lukeshu/src/github.com/envoyproxy/gateway/vendor/sigs.k8s.io/controller-runtime/pkg/client/fake/client.go:156 +0xcfb
  github.com/envoyproxy/gateway/internal/infrastructure/kubernetes.TestCreateOrUpdateServiceAccount.func1()
      /home/lukeshu/src/github.com/envoyproxy/gateway/internal/infrastructure/kubernetes/serviceaccount_test.go:90 +0x2cc
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/lib/go/src/testing/testing.go:1493 +0x47

Previous write at 0x00c0005b9930 by goroutine 39:
  k8s.io/apimachinery/pkg/apis/meta/v1.(*ObjectMeta).SetResourceVersion()
      /home/lukeshu/src/github.com/envoyproxy/gateway/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:146 +0x5c
  k8s.io/api/core/v1.(*ServiceAccount).SetResourceVersion()
      <autogenerated>:1 +0x29
  sigs.k8s.io/controller-runtime/pkg/client/fake.versionedTracker.Add()
      /home/lukeshu/src/github.com/envoyproxy/gateway/vendor/sigs.k8s.io/controller-runtime/pkg/client/fake/client.go:200 +0x1e6
  sigs.k8s.io/controller-runtime/pkg/client/fake.(*ClientBuilder).Build()
      /home/lukeshu/src/github.com/envoyproxy/gateway/vendor/sigs.k8s.io/controller-runtime/pkg/client/fake/client.go:156 +0xcfb
  github.com/envoyproxy/gateway/internal/infrastructure/kubernetes.TestCreateOrUpdateServiceAccount.func1()
      /home/lukeshu/src/github.com/envoyproxy/gateway/internal/infrastructure/kubernetes/serviceaccount_test.go:90 +0x2cc
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/lib/go/src/testing/testing.go:1493 +0x47

Goroutine 40 (running) created at:
  testing.(*T).Run()
      /usr/lib/go/src/testing/testing.go:1493 +0x75d
  github.com/envoyproxy/gateway/internal/infrastructure/kubernetes.TestCreateOrUpdateServiceAccount()
      /home/lukeshu/src/github.com/envoyproxy/gateway/internal/infrastructure/kubernetes/serviceaccount_test.go:83 +0x933
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/lib/go/src/testing/testing.go:1493 +0x47

Goroutine 39 (running) created at:
  testing.(*T).Run()
      /usr/lib/go/src/testing/testing.go:1493 +0x75d
  github.com/envoyproxy/gateway/internal/infrastructure/kubernetes.TestCreateOrUpdateServiceAccount()
      /home/lukeshu/src/github.com/envoyproxy/gateway/internal/infrastructure/kubernetes/serviceaccount_test.go:83 +0x933
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/lib/go/src/testing/testing.go:1493 +0x47
==================
```

2) `cmp` isn't actually the problem here. I'm guessing that changing out `cmp` just managed to change the timing enough that it doesn't trigger as often.  The problem is a few lines up, where it says?
```go
	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
```
All of those parallel tests share a single `tc` variable, that the loop is updating.
Fix that by changing it to
```go
	for _, tc := range testCases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
```
https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

**Review by LukeShu — CHANGES_REQUESTED** [review]
lgtm, other than the test race

File: internal/infrastructure/kubernetes/serviceaccount_test.go (hunk 4 above)
Comment (LukeShu) [reply internal/infrastructure/kubernetes/serviceaccount_test.go:96]: Oof, `git grep -B2 t.Parallel` finds a lot more instances of this mistake.

Comment (LukeShu) [reply internal/infrastructure/kubernetes/serviceaccount_test.go:96]: https://github.com/envoyproxy/gateway/pull/365

Comment (arkodg) [reply internal/infrastructure/kubernetes/serviceaccount_test.go]: thanks for digging deeper and fixing this throughout the codebase !
reverted the last commit that removed `cmp`

**Review by LukeShu — APPROVED** [review]
lgtm

**Pull request merged**